large-message-core/src/main/java/com/bakdata/kafka/LargeMessageRetrievingClient.java

```diff
@@ -46,38 +44,41 @@
 public class LargeMessageRetrievingClient {
 
     private final @NonNull Map<String, Supplier<BlobStorageClient>> clientFactories;
+    private final @NonNull HeaderLargeMessagePayloadSerde headerLargeMessagePayloadSerde;
     private final @NonNull Map<String, BlobStorageClient> clientCache = new HashMap<>();
 
-    static BlobStorageURI deserializeUri(final byte[] data) {
-        final byte[] uriBytes = getBytes(data);
-        final String rawUri = new String(uriBytes, CHARSET);
+    static BlobStorageURI deserializeUri(final byte[] uriBytes) {
+        final String rawUri = LargeMessagePayload.asUri(uriBytes);
         return BlobStorageURI.create(rawUri);
     }
 
-    static byte[] getBytes(final byte[] data) {
-        final byte[] bytes = new byte[data.length - 1];
-        // flag is stored in first byte
-        System.arraycopy(data, 1, bytes, 0, data.length - 1);
-        return bytes;
-    }
-
     /**
      * Retrieve a payload that may have been stored on blob storage
      *
      * @param data payload
+     * @param headers headers that might contain flag to distinguish blob storage backed messages
      * @return actual payload retrieved from blob storage
      */
-    public byte[] retrieveBytes(final byte[] data) {
+    public byte[] retrieveBytes(final byte[] data, final Headers headers) {
         if (data == null) {
             return null;
         }
-        if (data[0] == IS_NOT_BACKED) {
-            return getBytes(data);
+        final LargeMessagePayloadSerde serde = this.getSerde(headers);
```

Comment:
It depends on the headers whether we use the header version or not. This allows us to use both variants at the same time

Reply:
But for records without a flag it means that we have to iterate over all headers, right? In which situations do you even want to have both versions? I can't really estimate the performance impact of this, but maybe it would be worth it to make this configurable

Reply:
I don't think that iterating the header keys is expensive. Furthermore this allows us to ad-hoc switch to the new format without invalidating the old messages in a topic